ads/google/csa.js

```diff
@@ -0,0 +1,264 @@
+      }
+    }, 250); /* 250 is time in ms to wait before executing orientation */
+  });
+
+  // Only call for ads once the script has loaded
+  loadScript(global, 'https://www.google.com/adsense/search/ads.js', () => {
+    if (data['afsPageOptions'] != null && data['afshPageOptions'] == null) {
+      // Do not backfill this request
+      afsAdblockOptions['adLoadedCallback'] =
+          resizeIframe.bind(null, global, null, null);
+      // Make a call for AFS ads
+      global._googCsa('ads', afsPageOptions, afsAdblockOptions);
+    } else if (data['afsPageOptions'] == null
+          && data['afshPageOptions'] != null) {
+      // Do not backfill this request
+      afshAdblockOptions['adLoadedCallback'] =
+          resizeIframe.bind(null, global, null, null);
+      // Make a call for AFSh ads
+      global._googCsa('plas', afshPageOptions, afshAdblockOptions);
+    } else if (data['afsPageOptions'] != null
+          && data['afshPageOptions'] != null) {
+      // Do backfill this request
+      afshAdblockOptions['adLoadedCallback'] =
+          resizeIframe.bind(null, global, afsPageOptions, afsAdblockOptions);
+      // Make a call for AFSh ads (with AFS as backfill)
+      global._googCsa('plas', afshPageOptions, afshAdblockOptions);
+    }
+  });
+}
+
+/**
+ * CSA callback function to resize the iframe and/or request backfill
+ * @param {?Object} backfillPageOptions AFS page options (if necessary)
+ * @param {?Object} backfillAdblockOptions AFS ad unit options (if necessary)
+ * @param {string} containerName
+ * @param {boolean} adsLoaded
+ * @visibleForTesting
+ */
+export function resizeIframe(global, backfillPageOptions,
+    backfillAdblockOptions, containerName, adsLoaded) {
+  if (adsLoaded) {
+    // Get actual height of container
+    const container = global.document.getElementById(containerName);
+    const height = container.offsetHeight;
+    currentAmpHeight =
+        global.context.initialIntersection.boundingClientRect.height;
+    const overflowHeight = 40;
+
+    // If the height of the container is larger than the height of the
+    // initially requested AMP tag, add the overflow element
+    if (height > currentAmpHeight) {
+      // Create the overflow
+      createOverflow(global, overflowHeight, height, container,
+          currentAmpHeight - overflowHeight);
+    }
+
+    // Attempt to resize to actual CSA container height
+    global.context.requestResize(undefined, height);
+
+    // If reisze succeeded, hide overflow and resize container
+    global.context.onResizeSuccess(function(requestedHeight) {
+      currentAmpHeight = requestedHeight;
+      const overflow = global.document.getElementById('overflow');
+      if (overflow) {
+        overflow.style.display = 'none';
+        resizeCsa(container, requestedHeight);
+      }
+    });
+
+    // If resize was denied, show overflow element and resize container
+    global.context.onResizeDenied(function(requestedHeight) {
```

Comment:
Even though, I think logically it's better to move this logic out of the resize handler, to make sure we only register once.